[asl reference] LR1 parsing exposition (#1778)
herd/herdtools7: asllib/doc/Syntax.tex

\chapter{Syntax\label{chap:Syntax}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

This chapter defines the grammar of \ASLvOneterm{}.
The grammar is presented as a context-free grammar~\cite{HU79} with the help of
two shorthands --- \emph{inlined derivations} and \emph{parametric productions} ---
inspired  by the Menhir Parser Generator~\cite{MenhirManual} for the
\href{https://ocaml.org/}{OCaml language}.
These shorthands can be considered as macros over standard context-free grammar
notation, which can be expanded away to yield a standard context-free grammar.

\section*{Ambiguities and LR(1) grammars}

In most cases, \ASLvOneterm{}'s context-free grammar is unambiguous.
This means that for any given text, either:
\begin{itemize}
  \item It cannot be derived from the grammar. The text is therefore invalid
        ASL syntax, and causes a parse error. An example is \texttt{1 + * 3},
        which cannot be derived
        from the grammar in \secref{ASLGrammar} (intuitively, because \texttt{+}
        and \texttt{*} are binary operators that require two sub-expressions).

>         See \ExampleRef{Failing to Parse an Invalid Expression}.

Reviewer: ```suggestion
        ASL syntax, and causes a parse error.
```

  \item There is exactly one way to derive it from the grammar. The text is
        valid ASL syntax, and the derivation from the grammar describes how it
        should be bracketed.
\end{itemize}

> In some cases, there are ambiguities in \ASLvOneterm{}'s context-free grammar.

Reviewer: ```suggestion
In some cases, there seem to be ambiguities in \ASLvOneterm{}'s context-free grammar.
```

For example, the expression \texttt{1 + 2 * 3} can be derived with two different bracketings:
\begin{itemize}
  \item \texttt{(1 + 2) * 3}, which would evaluate to 9.
  \item \texttt{1 + (2 * 3)}, which would evaluate to 7.
\end{itemize}

To disambiguate these cases, \ASLvOneterm{} relies on the concept of an LR(1)
grammar~\cite{ASU86}.
In brief, an LR(1) grammar can be considered both a subcategory of context-free
grammars  and a specific kind of shift-reduce parsing algorithm.
Use of an LR(1) grammar ensures that the syntax of \ASLvOneterm{}:
\begin{itemize}
  \item Is demonstrably unambiguous.
  \item can be parsed deterministically in a single pass
        (that is, with no backtracking).
  \item can be concisely, precisely, and fully specified in both a
        machine-readable and formal way.
  \item permits use of a variety of parser generators to automatically produce
        a grammar-compliant parser from the machine-readable specification

>         of the grammar.

Reviewer: (consistency) these lines start with lower case, the others are upper case

        example in ASL, so multiplication is bracketed more tightly than addition).
  \item If the next token has lower priority, then reduce.
  \item Otherwise, the two tokens have the same priority, and the \shiftreduceconflictterm{} is resolved
        by assigning each priority an associativity:
        \begin{itemize}
        \item If the priority of the two tokens has left-associativity, reduce
        (favouring the left-associated bracketing).
        \item If the priority of the two tokens has right-associativity, shift
        (favouring the right-associated bracketing).
        \item Otherwise, the tokens are non-associative, and a parsing error is produced.
        \end{itemize}
  \end{itemize}
\end{itemize}

\ChapterOutline
\begin{itemize}
  \item \secref{InlinedDerivations} defines \emph{inlined derivations};
  \item \secref{ParametricProductions} defines \emph{parametric productions};
  \item \secref{ASLParametricProductions} introduces the parametric productions
        used for the ASL grammar;
  \item \secref{ASLGrammar} defines the ASL grammar;
  \item \secref{ParseTrees} defines parse trees;

>   \item \secref{PriorityAndAssociativity} defines the priority and associativity

Reviewer: We should rename this section to "Priority and Associativity for Parsing Tokens" or similar

\Lang{G} = \{\yield(n) \;|\; n \text{ is a well-formed parse tree for }G\} \enspace.
\]

\section{Priority and Associativity for Parsing Tokens\label{sec:PriorityAndAssociativity}}
As described in the introduction to \chapref{Syntax}, tokens that may cause a
\shiftreduceconflictterm{} are assigned a priority, and each priority is
assigned an associativity. Tokens that cannot cause \shiftreduceconflictsterm{}

> are not assigned priorities

Reviewer: Missing full stop

example \texttt{!!TRUE} still parses as \texttt{!(!TRUE)}.

\begin{center}
\begin{tabular}{lll}
\textbf{Priority level} & \textbf{Tokens} & \textbf{Associativity}\\
\hline
9 & $\Tdot$, $\Tlbracket$, $\Tllbracket$                                          & $\leftassoc$\\
8 & $\Tin$                                                                        & $\nonassoc$\\
7 & $\Tunops$                                                                     & $\nonassoc$\\
6 & $\Tpow$                                                                       & $\leftassoc$\\
5 & $\Tmul$, $\Tdiv$, $\Tdivrm$, $\Trdiv$, $\Tmod$, $\Tshl$, $\Tshr$              & $\leftassoc$\\
4 & $\Tplus$, $\Tminus$, $\Tor$, $\Tand$, $\Txor$, $\Tcoloncolon$, $\Tplusplus$   & $\leftassoc$\\
3 & $\Tgt$, $\Tgeq$, $\Tlt$, $\Tleq$                                              & $\nonassoc$\\
2 & $\Teqop$, $\Tneq$                                                             & $\leftassoc$\\
1 & $\Tbor$, $\Tband$, $\Tbimpl$, $\Tbeq$, $\Tas$                                 & $\leftassoc$\\
0 & $\Telse$                                                                      & $\nonassoc$
\end{tabular}
\end{center}

> \section{Parsing Examples\label{sec:ParsingExamples}}

Reviewer: Perhaps "Grammar Examples"?

This is how the priority of $\Tmul$ over $\Tplus$ is enforced.

The table above corresponds to the following derivation in the expanded grammar
(the underline indicates the non-terminal that is reduced at each step):
\[
\begin{array}{rl}
\Nexpr & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr}\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Nexpr \parsesep \Tmul \parsesep \underline{\Nexpr}\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Nexpr \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr} \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \underline{\Nexpr} \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \Tintlit(1) \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tmul \parsesep \Tintlit(3)\\
\end{array}
\]
Notice that the derivation corresponds perfectly to the concatenation of the stack
and input, read bottom-up in the table above, with the rightmost non-terminal
being the one that is reduced at each step.

> \ExampleDef{Shift-Reduce Ambiguity with Left Associativity}

Reviewer: "Resolving Ambiguity with Left Associativity"

\end{tabular}
\end{small}
\end{center}

As explained in the previous example, there are no reductions to $\Nunop$ or $\Nbinop$,
since they are inlined non-terminals.

\paragraph{Observation:}
At step 4, stack contains $\Tminus \parsesep \Nexpr$ and the next token is again $\Tminus$.
The parser reduces $\Tminus \parsesep \Nexpr$ to $\Nexpr$ instead of shifting the second $\Tminus$.
This is exactly the priority rule for unary minus:
the production $\Nexpr \derives \Tminus \parsesep \Nexpr$ has priority $\Tunops$,
which is higher than the binary-operator priority of $\Tminus$.
As a result, the input is parsed as $(-1) - 2$, not as $-(1 - 2)$.

The table above corresponds to the following derivation in the expanded grammar
(the underline indicates the non-terminal that is reduced at each step):
\[
\begin{array}{rl}
\Nexpr & \derives\ \underline{\Nexpr} \parsesep \Tminus \parsesep \Nexpr\\
       & \derives\ \underline{\Nexpr} \parsesep \Tminus \parsesep \Tintlit(2)\\
       & \derives\ \Tminus \parsesep \underline{\Nexpr} \parsesep \Tminus \parsesep \Tintlit(2)\\
       & \derives\ \Tminus \parsesep \Tintlit(1) \parsesep \Tminus \parsesep \Tintlit(2)\\
\end{array}
\]

> \ExampleDef{Parsing with Parametric Productions}

Reviewer: Remove "Parsing"

2 & $\Teqop$, $\Tneq$                                                             & $\leftassoc$\\
1 & $\Tbor$, $\Tband$, $\Tbimpl$, $\Tbeq$, $\Tas$                                 & $\leftassoc$\\
0 & $\Telse$                                                                      & $\nonassoc$
\end{tabular}
\end{center}

\section{Parsing Examples\label{sec:ParsingExamples}}

\ExampleDef{Derivation of a Valid Expression}
The expression \texttt{1 + 3}
is transformed by the lexical analysis into:
\[
\Tintlit(1) \parsesep \Tplus \parsesep \Tintlit(3)
\]

The corresponding derivation is:
\[
\begin{array}{rl}
\Nexpr & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr}\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \underline{\Nexpr} \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \Tintlit(1) \parsesep \Tplus \parsesep \Tintlit(3)\\
\end{array}
\]

This expression has the unique parse tree shown below.
\begin{center}
\begin{footnotesize}
\begin{adjustbox}{max width=\textwidth}
\Tree
[.$\Nexpr$
  [.$\Nexpr$ $\Tintlit(1)$ ]
  $\Tplus$
  [.$\Nexpr$ $\Tintlit(3)$ ]
]
\end{adjustbox}
\end{footnotesize}
\end{center}

> \ExampleDef{Failing to Parse an Invalid Expression}

Reviewer: ```suggestion
\ExampleDef{Failing to Derive an Invalid Expression}
```

The expression \texttt{1 + * 3}
is transformed by the lexical analysis into:
\[
\Tintlit(1) \parsesep \Tplus \parsesep \Tmul \parsesep \Tintlit(3)
\]

An attempted derivation gets stuck as follows:
\[
\begin{array}{rl}
\Nexpr & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr}\\
       & \not\derives\ \Nexpr \parsesep \Tplus \parsesep \Tmul \parsesep \Tintlit(3)
\end{array}
\]
because there is no derivation for $\Nexpr$ whose first token is $\Tmul$.

> \ExampleDef{Shift-Reduce Parsing with Binary Operators and Priority}

Reviewer: ```suggestion
\ExampleDef{Resolving Ambiguities with Binary Operators and Priority}
```

This is how the priority of $\Tmul$ over $\Tplus$ is enforced.

The table above corresponds to the following derivation in the expanded grammar
(the underline indicates the non-terminal that is reduced at each step):
\[
\begin{array}{rl}
\Nexpr & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr}\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Nexpr \parsesep \Tmul \parsesep \underline{\Nexpr}\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Nexpr \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr} \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \underline{\Nexpr} \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tmul \parsesep \Tintlit(3)\\
       & \derives\ \Tintlit(1) \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tmul \parsesep \Tintlit(3)\\
\end{array}
\]
Notice that the derivation corresponds perfectly to the concatenation of the stack
and input, read bottom-up in the table above, with the rightmost non-terminal
being the one that is reduced at each step.

> \ExampleDef{Shift-Reduce Parsing with Left Associativity}

Reviewer: ```suggestion
\ExampleDef{Resolving Ambiguity with Left Associativity}
```

to $\Nexpr$ instead of shifting the second $\Tplus$.
This is how the ambiguity is resolved using the fact that $\Tplus$
is declared $\leftassoc$:
when the competing actions have the same priority, left associativity chooses
reduce rather than shift.
As a result, the input is parsed as $(1 + 2) + 3$, not as $1 + (2 + 3)$.

The table above corresponds to the following derivation in the expanded grammar
(the underline indicates the non-terminal that is reduced at each step):
\[
\begin{array}{rl}
\Nexpr & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr}\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \underline{\Nexpr} \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \underline{\Nexpr} \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \Nexpr \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \underline{\Nexpr} \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tplus \parsesep \Tintlit(3)\\
       & \derives\ \Tintlit(1) \parsesep \Tplus \parsesep \Tintlit(2) \parsesep \Tplus \parsesep \Tintlit(3)\\
\end{array}
\]

> \ExampleDef{Shift-Reduce Parsing with Unary Operators}

Reviewer: ```suggestion
\ExampleDef{Derivation for Nested Unary Operators}
```
(or similar)

The expression \texttt{- - 1}
is transformed by the lexical analysis into:
\[
\Tminus \parsesep \Tminus \parsesep \Tintlit(1)
\]

The unique parse tree for this expression in the expanded grammar is:
\begin{center}
\begin{footnotesize}
\begin{adjustbox}{max width=\textwidth}
\Tree
[.$\Nexpr$
  $\Tminus$
  [.$\Nexpr$
    $\Tminus$
    [.$\Nexpr$ $\Tintlit(1)$ ]
  ]
]
\end{adjustbox}
\end{footnotesize}
\end{center}

> A corresponding shift-reduce parsing run is as follows:

Reviewer: I think the shift-reduce parsing run is unnecessary here - please delete

5 & \fbox{$\Tminus \parsesep \Tminus \parsesep \Nexpr$} $\emptysentence$ & reduce $\Nexpr$\\
6 & \fbox{$\Tminus \parsesep \Nexpr$} $\emptysentence$ & reduce $\Nexpr$\\
7 & \fbox{$\Nexpr$} $\emptysentence$ & accept
\end{tabular}
\end{small}
\end{center}

As explained in the previous examples, there is no reduction to $\Nunop$,
since it is an inlined non-terminal.

\paragraph{Observation:}
This example shows the parser building a nested unary expression.
After step 4, the stack suffix $\Tminus \parsesep \Tminus \parsesep \Nexpr$
matches the right-hand side obtained by expanding two applications of the unary-minus production.
The parser therefore performs two successive reductions of the form
$\Tminus \parsesep \Nexpr$ to $\Nexpr$,
which yields the parse $-(-1)$.

The table above corresponds to the following derivation in the expanded grammar
(the underline indicates the non-terminal that is reduced at each step):
\[
\begin{array}{rl}
\Nexpr & \derives\ \Tminus \parsesep \underline{\Nexpr}\\
       & \derives\ \Tminus \parsesep \Tminus \parsesep \underline{\Nexpr}\\
       & \derives\ \Tminus \parsesep \Tminus \parsesep \Tintlit(1)\\
\end{array}
\]

> \ExampleDef{Shift-Reduce Parsing with Unary Operators and Priority}

Reviewer: ```suggestion
\ExampleDef{Resolving Ambiguity with Unary Operators and Priority}
```
